doc/decisions/header_file_structure.md

```diff
@@ -12,31 +12,154 @@ The big problem is kdbprivate.h. It has two main problems:
 
 ## Constraints
 
+- Each library should have at least one separate header file:
+
+  Without this constraint, we would constantly include things that aren't needed.
+  This is inconvenient as it clutters the auto-suggestions of IDEs.
+  It also slows down the compile process, since the compiler has to parse everything that was included.
+
+- Non-public API and public API should be kept in separate files
+- Some non-public APIs need to be accessible for testing, such headers should not be packaged/installed
```

Comment:
What a coincidence! I just stumbled upon this while creating the high-level change-tracking API. Currently I just added a function for creating a fake/mock ChangeTrackingContext into the public API, but I thought maybe we should have seperate `-testing` headers for such functionality? This then would have to be installed, as 3rdpartys may write plugins that use it.

Reply:
> This then would have to be installed, as 3rdpartys may write plugins that use it.

Well, if the API is intended for consumption by third parties, then it should be installed anyway and not a separate header for testing. If people need more, they should just copy the headers from the repo into their code.